Utils/Blocking_queue/blocking_queue.h

```diff
@@ -0,0 +1,63 @@
+#ifndef BLOCKING_QUEUE_H
+#define BLOCKING_QUEUE_H
+#include <condition_variable>
+#include <queue>
+#include <iostream>
+
+template <typename T>
+class BlockingQueue{
+
+  std::mutex m_mutex;
+  std::condition_variable m_condition_variable;
+  std::queue<T> m_queue;
+  std::size_t m_size;
+
+  public:
+
+   explicit BlockingQueue(std::size_t size) : m_size(size) {}
+
+  void push(const T& item) noexcept{
+     std::unique_lock<std::mutex> lock(m_mutex);
+     m_condition_variable.wait(lock, [&]{return m_queue.size() < m_size;});
+     m_queue.push(item);
+     lock.unlock();
+     m_condition_variable.notify_one();
+   }
+
+   void push(T&& item) noexcept{
+     std::unique_lock<std::mutex> lock(m_mutex);
+     m_condition_variable.wait(lock, [&]{return m_queue.size() < m_size;});
+     m_queue.push(item);
+     lock.unlock();
+     m_condition_variable.notify_one();
+   }
+
+   bool try_push(T&& item) {
+     if(m_queue.size() < m_size) {
+       m_queue.push(item);
+     }
+     return m_queue.size() < m_size;
+   }
+
+   void pop() noexcept{
+     std::unique_lock<std::mutex> lock(m_mutex);
+     m_condition_variable.wait(lock, [&]{return !m_queue.empty();});
+     m_queue.pop();
+     lock.unlock();
+     m_condition_variable.notify_one();
+   }
+
+   bool try_pop() {
+     if(!m_queue.empty()) {
+       m_queue.pop();
+     }
+     return !m_queue.empty();
+   }
+
+   ~BlockingQueue() {
+     std::unique_lock<std::mutex> lock(m_mutex);
+     m_condition_variable.notify_all();
```

Comment:
Won't fix the issue. Suppose an empty queue, and multiple threads waiting to pop something out of there. You will notify them but nothing will change as those threads will sleep back because the queue is still empty.

Reply:
Your BlockingQueue destructor has a race condition that can lead to undefined behavior. What if BlockingQueue is destroyed, but some threads are still blocked inside pop().
When `m_condition_variable.notify_all()` is called, the threads wake up and try to access `m_merar_ara`, but the memory of the BlockingQueue may have already been freed.
Undefined behavior, potential segmentation fault, or memory corruption.